src/main/java/org/apache/maven/plugins/jar/AbstractJarMojo.java

```diff
@@ -76,7 +76,7 @@
     /**
      * The Jar archiver.
      */
-    @Component( role = Archiver.class, hint = "jar" )
+    @Component( role = Archiver.class, hint = "mjar" )
```

Comment:
Let's change this to

    @Component
    private Map<String, Archiver> archivers;

Now we have the option to decide which archiver to pick, e.g. based on the availability of the module descriptor

Reply:
The thing is how to know if the JAR file is going to contain `module-info.class` or not? The inclusion of the content is delegated to Maven Archiver:
```java
archiver.getArchiver().addDirectory( contentDirectory, getIncludes(), getExcludes() );
```
And in general the idea of `ModularJarArchiver` is to be safe to be used for "regular" JAR files as well. You can think for it as an `JarArchiver` that can handle modules if needed. The old `JarArchiver` could be used where you want to treat JAR strictly as archive - such as source JAR or if you explicitly want the module descriptor to be left unmodified (but is there such use case for the JAR plugin?)

That being said I agree that would be better if we know in advance if we need `ModularJarArchiver` or `JarArchiver` based on the existance of module descriptor (at very least will make the code intention more explicit and easy to read), so if there some acceptable way to know that I'll modify the PR.